---
title: "assignClustID: R package to find similarities between souporcell ouput and known genotypes"
author: Kristyna Kupkova
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true # table of content true
    toc_depth: 3  # upto three depths of headings (specified by #, ## and ###)
    theme: flatly # many options for theme, this one is my favorite.
    highlight: pygments
    df_print: paged
vignette: >
  %\VignetteIndexEntry{assignClustID: R package to find similarities between souporcell ouput and known genotypes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
editor_options: 
  markdown: 
    wrap: 72
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r klippy, echo=FALSE, include=TRUE}
# add copy clipboard button
klippy::klippy(position = "right")
```

```{r, echo=FALSE}
# These settings make the vignette prettier
knitr::opts_chunk$set(results="hold", collapse=FALSE, message=FALSE, 
                      warning = FALSE)
```

# Introduction to assignClustID

The assignClustID package purpose is to assign identity to clusters
produced by [souporcell](https://github.com/wheaton5/souporcell). The
functions provided within this package calculate and plot similarity
(proportion of exact matches) between cluster genotypes called with
souporcell and known genotypes.

# Setup

Let's load the assignClustID package and example datasets. There are two
inputs to `clustSimilarity`, which is the main function from this
package: 1) path to the VCF file generated by souporcell, 2)
*data.frame* with known genotypes. In the package we included example
VCF file generated by souporcell (*note*: this VCF file comes from a
poor-quality dataset, therefore we do not expect an extremely high
similarity between known phenotypes and the souporcell output), as well
as example *data.frame* (`DO_mm10`) of known genotypes. Lets' have a
look.

```{r setup}
library(assignClustID)

# 1) example VCF file - souporcell output
VCFpath = system.file("extdata", 
                      "cluster_genotypes.vcf", 
                      package = "assignClustID")

# show path
VCFpath
```

This is really just a full path to the VCF file including the file name.

```{r example_genotypes}
# 2) example data.frame with known genotypes
mm10_genotypes = DO_mm10

# show structure of know genotypes
head(mm10_genotypes)
```

As shown here the first three columns in the genotype *data.frame*
specify marker and it's genomic position. Following columns are genotype
IDs. If the genotype IDs are purely numerical, these should have an "X"
prefix in front of them (that's how they are going to be loaded).

**Side note:** The function `clustSimilarity` directly loads the VCF
file (using `fread` function from
[data.table](https://cran.r-project.org/web/packages/data.table/vignettes/datatable-intro.html)
package for speed). This is the format of the VCF file once it's loaded.

```{r example_VCF}
# show how internal functions loads a VCF file
VCFtable = assignClustID:::loadVCF(VCFpath)

head(VCFtable)
```

# Calculate similarities

Now that we have all the inputs, let's have a look how to use the
package functions.

First let's calculate similarities between the genotypes from the
souporcell output and all of the known genotypes.

```{r calculate_similarities}

similarities = clustSimilarity(pathToVCF=VCFpath, genotypes=mm10_genotypes)

# show output
similarities
```

The output of the `clustSimilarity` function is a table with calculated
similarities between known genotypes (columns) and souporcell clusters
(rows).

You can now plot the results with `plotClustSimilarity` function.

```{r plot_similarities, fig.width=5, fig.height=3.5}
simHeatmap = plotClustSimilarity(similarities)
```

To save the generated heatmap as pdf you can use following code.

```{r daveHeatmap, eval=FALSE}
pdf(file="heatmap.pdf", width = 4, height = 3.5)
draw(simHeatmap)
dev.off()
```

### Include only selected genotypes

There are few options included in the `clustSimilarity` function. For
example if you know IDs from known genotypes in your sample, you can use
just these IDs to calculate similarities.

```{r IDs}
includeIDs = paste0("X", c(110, 111, 120, 121))
includeIDs
```

```{r calculate_similarities_options, fig.width=4, fig.height=3.5}
selectSim= clustSimilarity(pathToVCF=VCFpath, 
                           genotypes=mm10_genotypes, 
                           IDs=includeIDs)

# plot and change palette to "RdBu"
simPlot = plotClustSimilarity(selectSim, RCBpalette="RdBu")
```

As you see, we are now missing comparisons to *X1* and *X150* genotypes.

